### PR TITLE
[Core]support external redis address when starting ray processes

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?
See RFC doc: https://docs.google.com/document/d/1hM6w4FYhfZVOP_CrWY9PNNPDuW_qe5SgPbCIQt3t_M8/edit?usp=sharing

## Usage
0. Assume we have a external redis server with endpoint `host:port` with password `xxx`
1. start head node with command: `ray start --head --address=host:port --redis-password="xxx"` (we reuse `--address` in parameter, which is forbidden in head node before)
2. start worker node with command: `ray start --address=host:port --redis-password="xxx"` (same as before)
3. connect program to ray cluster with `ray.init(address=host:port, _redis_password=xxx)` (same as before)

Note: If we have multiple external redis servers, set `--address` in step 1 as `host1:port1,host2:port2,...`
Limitation: If we have multiple redis servers, they should require same password.

## Related issue number

https://github.com/ray-project/ray/issues/13166

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Hey @WangTaoTheTonic @raulchen. 

Really great to see this change. But before we proceed more, can we create a simple API proposal in the Github issue? 

I think we'd like to eventually have the API that can be more generic right? (so that we can support other external storages other than Redis).

**Comment by WangTaoTheTonic** [comment]
Now we use `--external-redis-addresses` for specifying a external redis instance(s). What you mean by "more generic" is that we could use `--external-addresses` without wording about redis? (as I see the `--redis-address` has been changed to `--address`)

**Comment by rkooo567** [comment]
Yeah my suggestion is to just use `external-address`, but I am not sure if that'll be enough for more generic cases. One thing you can do is to create a RFC issue with the current proposal, and see what other people respond!

**Comment by WangTaoTheTonic** [comment]
ping @rkooo567 
A simplest version has been implemented. You can find the usage in the main post.
I believe there might has some details remains not handled. Comments are welcome and we can fix them now or quickly in future PRs!

File: python/ray/_private/parameter.py
Comment (raulchen) [inline python/ray/_private/parameter.py:15]: Better name it something like "external_gcs_storage_addresses". 
And we should document that this is addresses of the GCS's backend of storage service. Currently the storage is Redis, and we may have other options in the future.

Comment (WangTaoTheTonic) [reply python/ray/_private/parameter.py:15]: "external_gcs_storage_addresses" is more explicit than just `external_addresses`, and I like it.
@rkooo567 What do you think?

Comment (rkooo567) [reply python/ray/_private/parameter.py:15]: I think we don't want to have gcs here because it is the implementation detail (that was the primary reason why we used `address` instead of `gcs_address`). Btw, we are planning to remove Redis in the next quarter, and we should probably hold off this change. cc @ericl

Comment (WangTaoTheTonic) [reply python/ray/_private/parameter.py:15]: @rkooo567 Maybe you mixed up `gcs` with `redis`. Sure we DO NOT wanna include `redis` suffix/prefix in naming as implementation would drift in the future, but `gcs` is a SURE name in Ray. :-p

Comment (rkooo567) [reply python/ray/_private/parameter.py:15]: No. Gcs was the candidate for the api and it was rejected for the same reason

Comment (WangTaoTheTonic) [reply python/ray/_private/parameter.py:15]: Ohhh, didn't know about the gcs thing. Maybe we can change the name to "external_backend_storage_addresses" or just "external_storage_addresses"?

Comment (WangTaoTheTonic) [reply python/ray/_private/parameter.py:15]: or "external_metadata_storage_addresses"? Adding "metadata" to distinguish it from object store.

**Comment by rkooo567** [comment]
cc @wuisawesome

**Comment by wuisawesome** [comment]
Hey, looks like this PR has gotten lost. I don't think anyone is against adding functionality for external redis, we just need to get the API right (for example, we can probably just reuse `--address` instead of adding a new flag). 

@WangTaoTheTonic do you want to start an RFC issue for the API change?

**Comment by WangTaoTheTonic** [comment]
> Hey, looks like this PR has gotten lost. I don't think anyone is against adding functionality for external redis, we just need to get the API right (for example, we can probably just reuse `--address` instead of adding a new flag).
> 
> @WangTaoTheTonic do you want to start an RFC issue for the API change?

We've already started a RFC(listed in headline) and discussed about the parameter naming issue (resulted in `--external-addresses`.
The reason why we add an extra one instead of reuse `--address` is to tell head node the difference between using a remote storage or start a new one.
Let's push the functionality with former conclusion and not to discuss it again.

**Comment by wuisawesome** [comment]
I don't recall `--external-address` ever going through the core api approval process. cc @pcmoritz 

In any case, I still don't see what's wrong with using `--address`. One way to think about the behavior would be: we always connect to an external redis server. If one doesn't exist, then we fall back to creating one.

**Comment by WangTaoTheTonic** [comment]
> In any case, I still don't see what's wrong with using `--address`. One way to think about the behavior would be: we always connect to an external redis server. If one doesn't exist, then we fall back to creating one.

If you've read the codes you could find that `--address` is only used by raylets/workers but forbidden when starting head node.
Of course we can unlock this parameter in head node and treat it as `address of external redis`, then fall back to creating a new one with `--port` if we failed to connect to the `-address`.
This option doesn't introduce new api but changes the meaning of existing one. What do you think?

**Comment by wuisawesome** [comment]
That sounds like it makes sense to me, we need to get approval on it though. Btw what's the proposed behavior if we try to start multiple ray clusters with the same redis server?

**Comment by pcmoritz** [comment]
How about we do `--address=external://ip:port` and in that case it only connects and doesn't start redis. In that case the behavior is always clear and it can't be that some old redis that was lying around is re-used if the user used `--address=ip:port`.

**Comment by WangTaoTheTonic** [comment]
Now we don't support multiple clusters on one same redis server(even internally in Ant). It could be supported but lots of work need to be done as not only the storage part or the pub-sub part are designed for one cluster - one redis mode from the beginning.
We can support external redis first then see if it's necessary to support multiple clusters on the same redis server.
If reusing  `--address` is ok I would do changes first in this PR, is it ok? @wuisawesome @pcmoritz

**Comment by WangTaoTheTonic** [comment]
@wuisawesome @pcmoritz The code is ready as we discussed. Please take a look.
Even we have some variables named like `external-address` in implementation but those are only internal, not visible to users.

**Comment by WangTaoTheTonic** [comment]
ping @wuisawesome @pcmoritz

**Comment by WangTaoTheTonic** [comment]
@wuisawesome Hi Alex, do you have bandwidth to take a review?

**Review by wuisawesome — APPROVED** [review]

**Comment by zhe-thoughts** [comment]
Thanks for the contribution @WangTaoTheTonic

**Comment by yiwei00000** [comment]
Hi, does the current version of ray 1.4.1 not support ex_ redis？ @WangTaoTheTonic @rkooo567

**Comment by WangTaoTheTonic** [comment]
> Hi, does the current version of ray 1.4.1 not support ex_ redis？ @WangTaoTheTonic @rkooo567

I'm afraid the feature is not included in 1.4.1.

**Comment by yiwei00000** [comment]
> > Hi, does the current version of ray 1.4.1 not support ex_ redis？ @WangTaoTheTonic @rkooo567
> 
> I'm afraid the feature is not included in 1.4.1.

Will it be supported in the next version? When will the next version be released?

**Comment by WangTaoTheTonic** [comment]
> > > Hi, does the current version of ray 1.4.1 not support ex_ redis？ @WangTaoTheTonic @rkooo567
> > 
> > 
> > I'm afraid the feature is not included in 1.4.1.
> 
> Will it be supported in the next version? When will the next version be released?

I believe it will be included in 1.5.x version which would be released in next few days. The release manager is @jiaodong

**Comment by yiwei00000** [comment]
> > > > Hi, does the current version of ray 1.4.1 not support ex_ redis？ @WangTaoTheTonic @rkooo567
> > > 
> > > 
> > > I'm afraid the feature is not included in 1.4.1.
> > 
> > 
> > Will it be supported in the next version? When will the next version be released?
> 
> I believe it will be included in 1.5.x version which would be released in next few days. The release manager is @jiaodong

Thanks！！！

**Comment by jiaodong** [comment]
Hi, 1.5.0 branch is cut on last Monday 7/12 morning on fd59518980ba2995c72035042cc3341b75baef78, given this PR is merged 21 days ago it will be included once we're done with all tests :)

**Pull request merged**
